src/exchange_price_source.ts

```diff
@@ -0,0 +1,130 @@
+import { ExchangeAdapter, Ticker } from './exchange_adapters/base'
+import BigNumber from 'bignumber.js'
+import { PriceSource, WeightedPrice } from './price_source'
+import { Exchange, OracleCurrencyPair } from './utils'
+import { MetricCollector } from './metric_collector'
+import { individualTickerChecks } from './aggregator_functions'
+
+export interface OrientedExchangePair {
+  exchange: Exchange
+  symbol: OracleCurrencyPair
+  to_invert: boolean
+}
+
+export interface ExchangePriceSourceConfig {
+  pairs: OrientedExchangePair[]
+}
+
+export type OrientedAdapter = [ExchangeAdapter, boolean]
+
+export type PairData = {
+  bid: BigNumber
+  ask: BigNumber
+  baseVolume: BigNumber
+  quoteVolume: BigNumber
+}
+
+function invertPair(pair: PairData): PairData {
+  return {
+    bid: pair.ask.exponentiatedBy(-1),
+    ask: pair.bid.exponentiatedBy(-1),
+    baseVolume: pair.quoteVolume,
+    quoteVolume: pair.baseVolume,
+  }
+}
+
+function tickerToPairData(ticker: Ticker): PairData {
+  return {
+    bid: ticker.bid,
+    ask: ticker.ask,
+    baseVolume: ticker.baseVolume,
+    // Using lastPrice to convert from baseVolume to quoteVolume, whereas
+    // ideally one would use the baseVolume execution's VWAP.
+    quoteVolume: ticker.baseVolume.multipliedBy(ticker.lastPrice),
+  }
+}
+
+async function fetchPairData(
+  adapter: OrientedAdapter,
+  maxPercentageBidAskSpread: BigNumber,
+  metricCollector?: MetricCollector
+): Promise<PairData> {
+  const ticker = await adapter[0].fetchTicker()
+
+  // Validate fetched ticker -- will throw if invalid.
+  individualTickerChecks(ticker, maxPercentageBidAskSpread)
+
+  if (metricCollector) {
+    metricCollector.ticker(ticker)
+  }
+
+  const pair = tickerToPairData(ticker)
+  return adapter[1] ? invertPair(pair) : pair
+}
+
+function cumprod(array: BigNumber[]): BigNumber[] {
+  if (array.length == 0) {
+    return []
+  }
+  let sum: BigNumber[] = [array[0]]
```

Comment:
feels like sum may be a little misleading of a name?

Reply:
Indeed. Fixed.